### ISSUE TITLE
Decentralized Caching via IPFS or Torrent

### ISSUE THREAD

**Issue by chaxor** [snakemake/snakemake#2576]
In the standard GNU Make, the build system's efficiency lies within the nice DAG that ensures that a file is not re-created if it already exists.

A modern take on this may be that a git-commit's hash determines the files (and therefore the file hashes) that are built from the source code.  Therefore, it is obvious that for a given repo there is a hashmap of a gitcommit to a list of data-hashes.

It would be a wonderful feature to use this hashmap to ensure that very long running commands not be re-run during a build process.
In some sense, GNU's make does a first pass attempt at this problem, by simply checking for the local file.  However, tools such as ipfs or torrent today may eliminate some of these long running tasks even further, by looking up the hashes (guaranteed to be the correct output from this source-code build) online and downloading them directly rather than re-running long processes.

An example use case may be an ETL processes of generating a database, or generating exchange correlation functionals (each of which may take a few days on a supercomputer, but in some cases only be a few MBs or GBs).

I have been looking for this functionality for quite some time, as it should be standard in modern build-tools such as make.
Snakemake appears to be one of the few that is actively maintained and used compared to Make, and therefore I am interested in how to implement this functionality here.

**Describe the solution you'd like**
A very simple syntax addition (perhaps a prepended single character) to denote when a filepath should be checked on IPFS in addition to locally.
The idempotent logic should be something like:
  1) Is the filepath found locally?
      Yes) Goto (2) & return the file
      No) Goto (2)
   2) Is the ipfs_hash(filepath) on IPFS?
       Yes) Goto (4) & return the file
       No) Build the file and Goto (3)
   3) Upload the file to IPFS
   4) Download the file from IPFS to the filepath location

**Describe alternatives you've considered**
I have thought of doing this with Makefile, using something similar to this:
```ipfs_handle.sh
#!/bin/bash

file_path=$1
hash_file="$1.ipfs-hash"

# Check if the file exists
if [ -f "$file_path" ]; then
    # If file exists, check if it has an updated IPFS hash
    if [ -f "$hash_file" ]; then
        # File exists and has a known hash, use it as-is
        ipfs_hash=$(cat "$hash_file")
        echo "Using existing file with IPFS hash: $ipfs_hash"
    else
        # File exists but no known hash, upload to IPFS and store the hash
        new_hash=$(ipfs add -Q "$file_path")
        echo "$new_hash" > "$hash_file"
        echo "File uploaded to IPFS with new hash: $new_hash"
    fi
fi
```
And the the Makefile:
```Makefile
# Function to handle IPFS or local file
define ipfs_or_local
$(if $(findstring ipfs(,$1),\
$(eval file_path := $(shell echo '$1' | sed -e 's/ipfs(//g' -e 's/)//g'))\
$(shell ./ipfs_handle.sh $(file_path))\
$(file_path),\
$1)
endef

# Example target that generates a file
your-target: $(call ipfs_or_local,ipfs(path/to/generated/file))
    # Commands to generate the file
    <build-commands-for-your-file>
    # After build, handle IPFS upload and hash storage
    $(shell ./ipfs_handle.sh path/to/generated/file)
```

But it's not the nicest solution to be depending on across millions of repos, and would be better to be integrated into the build tool itself.

**Comment by Hocnonsense** [comment]
The `cached` function in snakemake may be helpful. However, it is mentioned but vaguely in the [document](https://snakemake.readthedocs.io/en/stable/snakefiles/writing_snakefiles.html#grammar).

**Comment by chaxor** [comment]
@Hocnonsense Yes, this `cached` function does seem useful, though after funding the [`storage` capabilities](https://snakemake.readthedocs.io/en/stable/snakefiles/storage.html), it seems that the best route is to create a storage plugin that works by calling and handling the ipfs protocol:

```Snakefile
rule example:
    input:
        storage("ipfs://{{cid}}/{{path}}")
    output:
        "example.txt"
    shell:
        "..."
```

This appears to be achievable to create this for downloading, although the automatic publishing and offering of the data to the community may be more challenging, as it is slightly different than the storage systems provided at the moment.  The ideal case for ipfs or torrent would be to automatically start or continue running a daemon which will add a node to the pool of servers containing the file for download.

The benefits of this are numerous and obvious:
  1) Data becomes more available with popularity
    - Likely surpassing the download speeds of centralized services
  2) Popular or useful data doesn't randomly disappear
    - Scientific data goes missing *constantly* due to researchers moving institutions (any university site download), or not continuing to pay huge fees to companies, or randomly deleting to open storage for personal files (most Google drive / Dropbox / AWS data disappears quickly in academia)
    3) Free
      - IPFS doesn't *require* paying any centralized company to host the data, but also doesn't restrict it
      
I don't have any experience with writing these snakemake plugins, but I would imagine getting the download functionality on ipfs or torrent would be fairly simple, while the real solution of having each client also have a server daemon with absolute minimal resources used might be slightly more tricky.
